tlsutil/config.go

```diff
@@ -915,22 +915,27 @@ func (c *Configurator) AuthorizeServerConn(dc string, conn *tls.Conn) error {
 	c.lock.RUnlock()
 
 	expected := c.ServerSNI(dc, "")
-	for _, chain := range conn.ConnectionState().VerifiedChains {
+	cs := conn.ConnectionState()
+	for _, chain := range cs.VerifiedChains {
 		if len(chain) == 0 {
 			continue
 		}
-		clientCert := chain[0]
-		_, err := clientCert.Verify(x509.VerifyOptions{
-			DNSName:   expected,
-			Roots:     caPool,
-			KeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
-		})
+		opts := x509.VerifyOptions{
+			DNSName:       expected,
+			Intermediates: x509.NewCertPool(),
+			Roots:         caPool,
+			KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+		}
+		for _, cert := range cs.PeerCertificates[1:] {
+			opts.Intermediates.AddCert(cert)
+		}
+		_, err := cs.PeerCertificates[0].Verify(opts)
 		if err == nil {
 			return nil
 		}
 		c.logger.Debug("AuthorizeServerConn failed certificate validation", "error", err)
```

Comment:
It might be nice to also set this log to emit at the Warning level.

Reply:
I added this log message originally because I was worried about exposing too much detail in the error message, but I realize now that all we do with that error is log it at the WARN level (in the caller).

Since this debug log line is in a loop, I think it might not be ideal to make this a warning, because if the client sends multiple cert chains we'll get warnings here that are misleading. 

To better expose the errors, maybe we should build up a slice of errors, and append those to the error message so that the caller can log all the details. If we did that we could probably remove this debug line because we'd have all the details in the returned error.